gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java

```diff
@@ -249,37 +248,37 @@ protected void executeCancellation() {
   /**
    * Add dependent jars and files.
    */
-  private void addDependencies()
+  private void addDependencies(Configuration conf)
       throws IOException {
     TimingEvent distributedCacheSetupTimer =
         this.eventSubmitter.getTimingEvent(TimingEvent.RunJobTimings.MR_DISTRIBUTED_CACHE_SETUP);
 
-    Path jarFileDir = new Path(this.mrJobDir, JARS_DIR_NAME);
+    Path jarFileDir = this.jarsDir;
```

Comment:
I think this is  dangerous as jarsDir will be shared across different jobs/runs. This may cause jar with mismatched versions to be stored. 

I think you have two options:
- Use both a job-specific and global directory. Use the latter if the name and size match, otherwise upload to the job-specific dir.
- Have a more complex layout for the jar cache which allows different versions to be stored (like the gradle cache).